buildSrc/src/main/kotlin/org/gradle/plugins/buildscan/BuildScanConfigurationPlugin.kt

```diff
@@ -120,81 +127,79 @@ class BuildScanConfigurationPlugin : Plugin<Project> {
+                .split(",")
 
-            project.buildScan.buildFinished({
-                allprojects.flatMap { it.tasks }.forEach {
-                    if (it.state.executed && (it.path in tasksToInvestigate)) {
+            buildScan.buildFinished {
+                allprojects.flatMap { it.tasks }
+                    .filter { it.state.executed && it.path in tasksToInvestigate }
+                    .forEach { task ->
                         val hasher = (gradle as GradleInternal).services.get(ClassLoaderHierarchyHasher::class.java)
-                        Visitor(project.buildScan, hasher, it).visit(it::class.java.classLoader)
+                        Visitor(buildScan, hasher, task).visit(task::class.java.classLoader)
                     }
-                }
-            })
+            }
         }
     }
 
-    private fun Project.extractAllReportsFromCI(project: Project) {
+    private
+    fun Project.extractAllReportsFromCI() {
         val capturedReportingTypes = listOf("html") // can add xml, text, junitXml if wanted
         val basePath = "${System.getenv("BUILD_SERVER_URL")}/repository/download/${System.getenv("BUILD_TYPE_ID")}/${System.getenv("BUILD_ID")}:id"
 
         gradle.taskGraph.afterTask {
-            if (state.failure != null) {
-                if (this is Reporting<*>) {
-                    val reportContainer = this as ReportContainer<*>
-                    reportContainer
-                        .filter { it.name in capturedReportingTypes && it.isEnabled && it.destination.exists() }
-                        .forEach {
-                            val linkName = "${this::class.java.simpleName.split("_")[0]} Report ($path)" // Strip off '_Decorated' addition to class names
-                            // see: ciReporting.gradle
-                            val reportPath = if (it.destination.isDirectory) {
-                                "report-${project.name}-${it.destination.name}.zip"
-                            } else {
-                                "report-${project.name}-${it.destination.parentFile.name}-${it.destination.name}"
-                            }
-                            val reportLink = "$basePath/$reportPath"
-                            project.buildScan.link(linkName, reportLink)
-                        }
-                }
+            if (state.failure != null && this is Reporting<*>) {
+                this.reports.filter { it.name in capturedReportingTypes && it.isEnabled && it.destination.exists() }
+                    .forEach { report ->
+                        val linkName = "${this::class.java.simpleName.split("_")[0]} Report ($path)" // Strip off '_Decorated' addition to class names
+                        // see: ciReporting.gradle
+                        val reportPath =
+                            if (report.destination.isDirectory) "report-${project.name}-${report.destination.name}.zip"
```

Comment:
I guess we should somehow re-use the logic from `CiReportingPlugin` here, right.
Even better: Let's add this logic to the `CiReportingPlugin`, so it can add the extra data to the build scan in a `.withPlugin` block.
WDYT?